File: data/solutions/fencing_naive.py
R, C = map(int, input().split())

grid = tuple(tuple(map(int, input().split())) for r in range(R))

maxPerim = 0

for r in range(R):
    for c in range(C):
        maxPerim = max(maxPerim, grid[r][c])
        for h in range(2, R - r + 1):
            for w in range(2, C - c + 1):
                try:
                    perim = 0
                    # horizontal
                    perim += sum(grid[r][c:c+w])
                    perim += sum(grid[r+h-1][c:c+w])
                    # vertical
                    perim += sum([ grid[r2][c] for r2 in range(r+1, r+h-1) ])
                    perim += sum([ grid[r2][c + w - 1] for r2 in range(r+1, r+h-1) ])
                    # update max
                    maxPerim = max(perim, maxPerim)
                except IndexError as e:
                    print(r, c, w, h)
                    raise e

print(maxPerim)

File: data/solutions/fencing_naive.c
#include <stdio.h>
#include <stdlib.h>

#define max(x,y) x>y?x:y

int ans(int R, int C, int** grid) {
  int maxPerim = grid[0][0];

  for (int r = 0; r < R; r++) {
    for (int c = 0; c < C; c++) {
      maxPerim = max(maxPerim, grid[r][c]);

      for (int h = 2; h < R - r + 1; h++) {
        for (int w = 2; w < C - c + 1; w++) {
          int perim = 0;
          // horizontal
          for (int d = c; d < c+w; d++) {
            perim += grid[r][d];
            perim += grid[r+h-1][d];
          }

          // vertical
          for (int d = r+1; d < r+h-1; d++) {
            perim += grid[d][c];
            perim += grid[d][c + w - 1];
          }

          // update max
          maxPerim = max(perim, maxPerim);
        }
      }
    }
  }
  return maxPerim;
}

int main() {
  int R = 0, C = 0;
  scanf("%d %d", &R, &C);

  int** grid = malloc(R * sizeof(int*));

  for (int r = 0; r < R; r++) {
    grid[r] = malloc(C * sizeof(int));
    for (int c = 0; c < C; c++) {
      scanf("%d", &grid[r][c]);
    }
  }

  int maxPerim = ans(R, C, grid);

  for (int r = 0; r < R; r++) {
    free(grid[r]);
  }
  free(grid);

  printf("%d", maxPerim);
  return 0;
}
